File: TUnit.TestProject/InjectedClassDataSourceWithAsyncInitializerTests.cs
using TUnit.Core.Interfaces;
#pragma warning disable CS9113 // Parameter is unread.

namespace TUnit.TestProject;

[ClassDataSource<MyClass>(Shared = SharedType.Keyed, Key = "MyKey")]
public class InjectedClassDataSourceWithAsyncInitializerTests(InjectedClassDataSourceWithAsyncInitializerTests.MyClass myClass)
{
    [Before(Test)]
    public Task BeforeTest()
    {
        Console.WriteLine("BeforeTest");
        return Task.CompletedTask;
    }

    [Test]
    public Task Test1()
    {
        Console.WriteLine("Test");
        return Task.CompletedTask;
    }
    
    [Test]
    public Task Test2()
    {
        Console.WriteLine("Test");
        return Task.CompletedTask;
    }
    
    [Test]
    public Task Test3()
    {
        Console.WriteLine("Test");
        return Task.CompletedTask;
    }

    public class MyClass : IAsyncInitializer
    {
        public Task InitializeAsync()
        {
            Console.WriteLine("IAsyncInitializer.InitializeAsync");
            return Task.CompletedTask;
        }
    }
}

File: TUnit.TestProject.FSharp/InjectedClassDataSourceWithAsyncInitializerTests.fs
namespace TUnit.TestProject.FSharp

open System
open System.Threading.Tasks
open TUnit.Core
open TUnit.Core.Interfaces


type MyClass() =
        interface IAsyncInitializer with
            member _.InitializeAsync() =
                Console.WriteLine("IAsyncInitializer.InitializeAsync")
                Task.CompletedTask

[<ClassDataSource(typeof<MyClass>, Shared=SharedType.Keyed, Key="MyKey")>]
type InjectedClassDataSourceWithAsyncInitializerTests(myClass: MyClass) =
    [<Before(HookType.Test)>]
    member _.BeforeTest() : Task =
        Console.WriteLine("BeforeTest")
        Task.CompletedTask

    [<Test>]
    member _.Test1() : Task =
        Console.WriteLine("Test")
        Task.CompletedTask

    [<Test>]
    member _.Test2() : Task =
        Console.WriteLine("Test")
        Task.CompletedTask

    [<Test>]
    member _.Test3() : Task =
        Console.WriteLine("Test")
        Task.CompletedTask
